src/java/org/apache/cassandra/index/sai/utils/IndexTermType.java

```diff
@@ -525,11 +525,15 @@ public ByteSource asComparableBytes(ByteBuffer value, ByteComparable.Version ver
     {
         if (isInetAddress() || isBigInteger() || isBigDecimal())
             return ByteSource.optionalFixedLength(ByteBufferAccessor.instance, value);
+        else if (isLong())
             // The LongType.asComparableBytes uses variableLengthInteger which doesn't play well with
             // the balanced tree because it is expecting fixed length data. So for SAI we use a optionalSignedFixedLengthNumber
             // to keep all comparable values the same length
-        else if (isLong())
             return ByteSource.optionalSignedFixedLengthNumber(ByteBufferAccessor.instance, value);
+        else if (isFrozen())
+            // We need to override the default frozen implementation here because it will defer to the underlying
+            // type's implementation which will be incorrect, for us, for the case of multi-cell types.
+            return ByteSource.of(value, version);
```

Comment:
Trying to clarify here...

If the underlying implementation is incorrect for multi-cell types, why is it the frozen case that we override?

Reply:
It's really because we don't support proper indexing on frozen indexing, so we don't want the underlying collection as a multi-component `ByteComparable` but as a single-component `ByteComparable`. I'll try and come up with some wording here that is clearer. We could even mention the ticket for full indexing on frozen collections in the comment.